crates/matrix-sdk-crypto/src/identities/user.rs

```diff
@@ -726,12 +740,16 @@ impl ReadOnlyUserIdentity {
         master_key: MasterPubkey,
         self_signing_key: SelfSigningPubkey,
     ) -> Result<(), SignatureError> {
-        master_key.verify_subkey(&self_signing_key)?;
+        if !self_signing_key.usage().contains(&KeyUsage::SelfSigning) {
```

Comment:
Where do we check that `master_key` has the correct usage?

Reply:
Ah, so it's in `verify_subkey`. Might make sense to consider moving the check to `MasterPubkey` construction time, so that we cannot even construct it from a `CrossSigningKey` with an incorrect usage. Same for the other two key types.

Reply:
We now directly deserialize into the per-usage specific type: https://github.com/matrix-org/matrix-rust-sdk/pull/1362/commits/6cb1e8a156966847031706cadc9e0896887e31fc.

Reply:
Great, thanks!